### PR TITLE
Error building count statement in pageExecute() function when base SQL includes sub-selects, see #715

### PR DESCRIPTION
**Pull request by mnewnham**
1. Complex SQL statements cause the statement rewrite in pageExecute() to fail.
2. Add support for the PDO/mysql driver in the pageExecute() function

### CODE REVIEW

File: adodb-lib.inc.php [hunk 1]
```diff
@@ -408,22 +423,32 @@ function _adodb_getcount(&$zthis, $sql,$inputarr=false,$secs2cache=0)
...
 		if ( strpos($sql, '_ADODB_COUNT') !== FALSE ) {
 			$rewritesql = preg_replace('/^\s*?SELECT\s+_ADODB_COUNT(.*)_ADODB_COUNT\s/is','SELECT COUNT(*) ',$sql);
 		} else {
+			
+			$ssMatches = false;
+			/*
+			* subselect will either end in , or from, which will 
+			* include either an alias or not
+			*/
+			$subSelect = preg_match_all ('/(\(SELECT\s.*\s+FROM\s.*\)\s.*\w.*\s)(,|FROM)/i',$sql,$ssMatches);
```
[the remark this answers is not in the document]
Comment (mnewnham) [reply adodb-lib.inc.php:440]: Thanks for intervening on this, for I am shit at regexps. It's not quite like that, the incoming statement might look like this

````
SELECT a,b, (SELECT MAX(fld) FROM x WHERE blah,blah) alias1 FROM y WHERE......
````
remembering of course that:
1. The selected fields can be in any order
2. There may be multiple subselects

The idea is for all the subselects to be eliminated so we can end up with

`````
SELECT COUNT(*) FROM (SELECT a,b FROM y WHERE.....) _ADODB_ALIAS_
````

**Comment by dregad** [comment]
@mnewnham the more I think about it, the less I'm sure we can reliably covering all cases using regex.

I tried with a recursive regex approach (e.g. `\((?:(?>[^()]+)|(?R))*\)\s+\w+\s*,?\s*` , which correctly identifies all subqueries (even if they have subqueries of their own), but the problem is that there can also be subqueries after the main FROM clause, which we must keep, but those are matched and removed too, and I can't think of a way to reliably isolate them at the moment.

This may be the reason why the `_ADODB_COUNT` hack was introduced (see #88).

Since what we want to do is simply replace everything between SELECT and FROM by COUNT(*), we could use a different approach, like parsing the query string character-by-character until we get to the first FROM that is not inside parentheses 

I'm also wondering why we don't use the `SELECT COUNT(*) FROM (" . $rewritesql . ")"` approach used  everywhere ? There is a comment stating _this is only supported by oracle and postgresql_, but I just tried with mysqli and it works just fine...

**Comment by mnewnham** [comment]
I think it's important to remember that many of these comments were added 20 years ago, when databases like mysql were shadows of what they are today. There are many cases here where complete refactoring is the appropriate response. 
In this case, I think it makes perfect sense to go with the "oracle only" solution for everything. I think that this will simplify this tremendously.

Incidentally, if you ever need to try out a solution for cross-compatibility, I've got ADOdb systems set up for Oracle, SQL Server(Windows and Linux), DB2, Postgres, MySQL (Both MariaDB and MySQL 8), firebird, odbc, and access. I've even got old systems for informix and SAP MaxDB so I can easily grab code and check it.

**Comment by dregad** [comment]
Right, that makes sense.

However, considering that this is targeted at 5.21.1, maybe we should have a 2-step approach of first applying a targeted fix (getting rid of regex logic), and target the bigger refactoring at 5.22 to avoid introducing regressions ?

**Comment by dregad** [comment]
@mnewnham Following up on my earlier post, please review #723 as proposed "step 1".

**Comment by dregad** [comment]
Removed tags and milestones, as they were duplicating metadata already set on original issue #715

**Comment by dregad** [comment]
Closing this in favor of #723.

**Pull request closed without merging**